torch/distributed/_spmd/graph_optimization.py

```diff
@@ -548,3 +554,339 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+
+    def replace_flatten_output_args(orig_node: fx.Node, new_node: fx.Node):
+        for idx in flatten_output_args_indices[orig_node]:
+            flatten_output_args[idx] = new_node
+
+    # Create the new step and optim nodes and blocks.
+    for group_idx in range(2):
+        step_args: List[fx.Node] = []
+        orig_step_outputs: List[fx.Node] = []
+        # We have to create the new step node and block first because it is used
+        # for the new optim node as the input.
+        with gm.graph.inserting_after(orig_optim_block.optim.optim_node):
+            for idx in orig_step_indices[group_idx]:
+                step_args.append(
+                    cast(Tuple[fx.Node, ...], orig_optim_block.step.add_node.args[0])[
+                        idx
+                    ]
+                )
+                orig_step_outputs.append(orig_optim_block.step.outputs[idx])
+            step = gm.graph.call_function(
+                torch.ops.aten._foreach_add.Scalar,
+                (step_args, 1),
+            )
+        step_block = ForeachAddBlock(step, generate_output=True)
+        for i, step_output in enumerate(step_block.outputs):
+            # Replace the original step output in the graph output node with
+            # the new one.
+            orig_step_output = orig_step_outputs[i]
+            replace_flatten_output_args(orig_step_output, step_output)
+            # Also need to replace the step output used for the new optimizer.
+            assert optim_args[group_idx].state_steps[i] == orig_step_output, (
+                f"The expected step output node mismatched, {orig_step_output} "
+                f"{optim_args[group_idx].state_steps[i]}"
+            )
+            optim_args[group_idx].state_steps[i] = step_output
+
+        # Insert the optimizer node after the first step output because its
+        # topo sort order is the last.
+        with gm.graph.inserting_after(step_block.outputs[0]):
+            optim = gm.graph.call_function(
+                torch.ops.aten._fused_adam.default,
+                optim_args[group_idx],
+                orig_optim_block.optim.optim_node.kwargs,
+            )
+        optim_block = FusedAdamBlock(optim, generate_output=True)
+        for curr_idx, orig_idx in enumerate(orig_optim_indices[group_idx]):
+            list_names = ("param_outputs", "exp_avgs_outputs", "exp_avg_sqs_outputs")
+            for name in list_names:
+                orig_list = getattr(orig_optim_block.optim, name)
+                curr_list = getattr(optim_block, name)
+                replace_flatten_output_args(orig_list[orig_idx], curr_list[curr_idx])
+
+        result.append(FusedOptimizerBlock(step_block, optim_block))
+
+    # Optimizer is used as the output of the train_step. Therefore, we have to
+    # update the output node of the graph.
+    output_args, output_kwargs = tree_unflatten(flatten_output_args, spec)
+    gm.graph.node_set_args(output, output_args)
```

Comment:
nit: a little bit inconsistent with the convention in `iter_graph_module.py`, where we directly do `output.args = output_args` there